src/transformers/core_model_loading.py

```diff
@@ -260,37 +332,111 @@ class WeightConverter:
     - target_keys: str | list[str] | None
     - distributed_operation / operations / quantization_operations are ALWAYS lists.
 
-    TODO: for BNB we need to collect model.weight.quant_state_keys
     """
 
-    source_keys: Union[str, list[str]]
-    target_keys: Optional[Union[str, list[str]]] = None
+    source_keys: Union[str, list[str]] = field(init=True)
+    target_keys: Union[str, list[str]] = field(init=True)
     operations: list[ConversionOps] = field(default_factory=list, repr=False)
 
     distributed_operation: Optional[TensorParallelLayer] = None
     quantization_operation: Optional[ConversionOps] = None
 
+    collected_tensors: dict[str, list[Future]] = field(default_factory=dict, init=False)
+    layer_targets: dict[str, set[str]] = field(default_factory=dict, init=False)
+
     def __post_init__(self):
-        if not isinstance(self.source_keys, list):
+        if isinstance(self.source_keys, str):
             self.source_keys = [self.source_keys]
-        targets_were_none = False
-        if not isinstance(self.target_keys, list):
-            if self.target_keys is None:
-                self.target_keys = list(self.source_keys)
-                targets_were_none = True
-            else:
-                self.target_keys = [self.target_keys]
+        if isinstance(self.target_keys, str):
+            self.target_keys = [self.target_keys]
 
-        if not targets_were_none and bool(len(self.source_keys) - 1) + bool(len(self.target_keys) - 1) >= 2:
+        if bool(len(self.source_keys) - 1) + bool(len(self.target_keys) - 1) >= 2:
             raise ValueError(
                 f"source keys={self.source_keys}, target_keys={self.target_keys} but you can only have one to many, one to one or many to one."
             )
 
+        if not self.operations:
+            raise ValueError("WeightConverter requires at least one operation.")
+
+    def add_tensor(self, target_key: str, original_key: str, source_pattern: str, future: Future):
```

Comment:
```suggestion

    def add_tensor(self, target_key: str, source_key: str, source_pattern: str, future: Future):
```
Let's always talk about sources and targets, instead of mixing names!